ASSERTION m_repaintContainer == repaintContainer on reddit.com

mattwoodrow opened the pull request:
#### b755716f3cbfd8eceb3908a57b06eb5d24469767
<pre>
ASSERTION m_repaintContainer == repaintContainer on reddit.com
<a href="https://bugs.webkit.org/show_bug.cgi?id=289138">https://bugs.webkit.org/show_bug.cgi?id=289138</a>
&lt;<a href="https://rdar.apple.com/146186159">rdar://146186159</a>&gt;

Reviewed by Simon Fraser.

This happens because ending a backing sharing sequence updates the repaint rects
(and repaint containers) of all the layers in the sequence.  In some cases we
decide a layer will join a sequence, a descendant of that layer causes the end
of the sequence, and then we make a post-descendant decision to composite the
layer (and stop participating in the already ended sharing sequence).  In this
case the repaint container becomes the sharing layer, and we don&apos;t change it
back again to be itself.

This adds new calls to updateLayerPositions in debug builds, which verify that
the whole tree is in a valid state during compositing, and causes existing tests
to catch this bug. Also adds some extra details in the layer tree debug log, in
order to determine what the repaint container for a layer should be.

The fix is to stop updating repaint rects when we finish a sequence, but instead
compute them when we finish processing a layer (as we already for
composited/not-composited changes).

The hard bit is that the before/after state is hard to determine, since a layer
might have its backing sharing layer removed before we even reach it (if a new
layer was inserted earlier that introduces overlap and ends the sequence), or
might not be added to a sharing layer until after processing (if the sequence it
participates in continues for latter siblings). Adds a new CheckedPtr to
RenderLayer to track the &apos;previous&apos; state to get around this, which is fairly
unfortunate.

Also adds flags to all the functions that change backing sharing state, so that
mutations from a style change have the repaints applied immediately, whereas
changes during a compositing update get their repaints deferred until the end of
the processing the relevant layer.

The only alternative approach I thought of was to defer backing sharing changes
during a compositing update, instead building a new datastructure of proposed
changes (like BackingSharingState, but accumulating through the entire update).
At the end of updating compositing layer, do a post-process to apply any sharing
changes (adding or removing).

* LayoutTests/platform/mac-wk1/compositing/repaint/composite-and-scale-change-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/fast/repaint/translate-animation-repaint-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/svg/compositing/transform-change-repainting-no-viewBox-repaintRects-expected.txt:
* LayoutTests/platform/mac-wk1/svg/compositing/transform-change-repainting-viewBox-repaintRects-expected.txt:
* LayoutTests/svg/compositing/transform-change-repainting-viewBox-repaintRects-expected.txt:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::~RenderLayer):
(WebCore::RenderLayer::removeSelfFromCompositor):
(WebCore::RenderLayer::recursiveUpdateLayerPositions):
(WebCore::RenderLayer::computeRepaintRects):
(WebCore::RenderLayer::setBackingProviderLayer):
(WebCore::RenderLayer::disconnectFromBackingProviderLayer):
(WebCore::RenderLayer::calculateClipRects const):
(WebCore::outputLayerPositionTreeLegend):
(WebCore::outputLayerPositionTreeRecursive):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::willBeDestroyed):
(WebCore::clearBackingSharingLayerProviders):
(WebCore::RenderLayerBacking::setBackingSharingLayers):
(WebCore::RenderLayerBacking::removeBackingSharingLayer):
(WebCore::RenderLayerBacking::clearBackingSharingLayers):
* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::BackingSharingState::issuePendingRepaints):
(WebCore::RenderLayerCompositor::updateCompositingLayers):
(WebCore::RenderLayerCompositor::computeCompositingRequirements):
(WebCore::RenderLayerCompositor::updateBackingSharingBeforeDescendantTraversal):
(WebCore::RenderLayerCompositor::updateBackingSharingAfterDescendantTraversal):
(WebCore::RenderLayerCompositor::layerGainedCompositedScrollableOverflow):
(WebCore::RenderLayerCompositor::layerStyleChanged):
(WebCore::clearBackingSharingWithinStackingContext):
(WebCore::RenderLayerCompositor::updateExplicitBacking):
(WebCore::RenderLayerCompositor::updateBacking):
(WebCore::RenderLayerCompositor::updateReflectionCompositingState):
(WebCore::RenderLayerCompositor::repaintOnCompositingChange):
(WebCore::RenderLayerCompositor::layerWillBeRemoved):
(WebCore::RenderLayerCompositor::clearBackingForAllLayers):
(WebCore::RenderLayerCompositor::updateLayerCompositingState): Deleted.
* Source/WebCore/rendering/RenderLayerCompositor.h:

Canonical link: <a href="https://commits.webkit.org/292627@main">https://commits.webkit.org/292627@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/631bf0b500a7bfb141cf7936d471a81ab40ff8fd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96642 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16256 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/6365 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101714 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/47162 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/98687 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16552 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24691 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/73646 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/47162 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99645 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/12450 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/87388 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/53982 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/12209 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5184 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46490 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/82310 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/5274 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103738 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23709 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/17281 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/82697 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23960 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83437 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/82077 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20604 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26728 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/4255 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/17183 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23672 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/28827 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/23331 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/26811 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25072 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderLayer.h:
```diff
@@ -770,7 +774,7 @@ class RenderLayer final : public CanMakeSingleThreadWeakPtr<RenderLayer>, public
     LayoutRect repaintRectIncludingNonCompositingDescendants() const;
 
-    void setRepaintStatus(RepaintStatus status) { m_repaintStatus = status; }
+    void setRepaintStatus(RepaintStatus status) { m_repaintStatus = status; setNeedsPositionUpdate(); }
```
I'd appreciate a comment to say why changing the repaint status implies we need to call `setNeedsPositionUpdate()`. Should you check that the status changed?

smfr commented on Source/WebCore/rendering/RenderLayer.h:
```diff
@@ -855,14 +859,18 @@ class RenderLayer final : public CanMakeSingleThreadWeakPtr<RenderLayer>, public
 
     bool paintsIntoProvidedBacking() const { return !!m_backingProviderLayer; }
 
+    RenderLayer* backingProviderLayerAtEndOfCompositingUpdate() const { return m_backingProviderLayerAtEndOfCompositingUpdate.get(); }
```
Is this used outside of assertions? If not, #ifdef it.

smfr commented on Source/WebCore/rendering/RenderLayer.h:
```diff
@@ -1380,10 +1388,8 @@ class RenderLayer final : public CanMakeSingleThreadWeakPtr<RenderLayer>, public
-#if ASSERT_ENABLED || ENABLE(CONJECTURE_ASSERT)
-    SingleThreadWeakPtr<RenderObject> m_repaintContainer;
-#endif
+    SingleThreadWeakPtr<RenderLayer> m_backingProviderLayerAtEndOfCompositingUpdate;
```
Is this used outside of assertions? If not, #ifdef it.

mattwoodrow replied:
It's used as part of the actual logic (which I am slightly sad about).

mattwoodrow committed:
ASSERTION m_repaintContainer == repaintContainer on reddit.com
https://bugs.webkit.org/show_bug.cgi?id=289138
<rdar://146186159>

Reviewed by Simon Fraser.

This happens because ending a backing sharing sequence updates the repaint rects
(and repaint containers) of all the layers in the sequence.  In some cases we
decide a layer will join a sequence, a descendant of that layer causes the end
of the sequence, and then we make a post-descendant decision to composite the
layer (and stop participating in the already ended sharing sequence).  In this
case the repaint container becomes the sharing layer, and we don't change it
back again to be itself.

This adds new calls to updateLayerPositions in debug builds, which verify that
the whole tree is in a valid state during compositing, and causes existing tests
to catch this bug. Also adds some extra details in the layer tree debug log, in
order to determine what the repaint container for a layer should be.

The fix is to stop updating repaint rects when we finish a sequence, but instead
compute them when we finish processing a layer (as we already for
composited/not-composited changes).

The hard bit is that the before/after state is hard to determine, since a layer
might have its backing sharing layer removed before we even reach it (if a new
layer was inserted earlier that introduces overlap and ends the sequence), or
might not be added to a sharing layer until after processing (if the sequence it
participates in continues for latter siblings). Adds a new CheckedPtr to
RenderLayer to track the 'previous' state to get around this, which is fairly
unfortunate.

Also adds flags to all the functions that change backing sharing state, so that
mutations from a style change have the repaints applied immediately, whereas
changes during a compositing update get their repaints deferred until the end of
the processing the relevant layer.

The only alternative approach I thought of was to defer backing sharing changes
during a compositing update, instead building a new datastructure of proposed
changes (like BackingSharingState, but accumulating through the entire update).
At the end of updating compositing layer, do a post-process to apply any sharing
changes (adding or removing).

* LayoutTests/platform/mac-wk1/compositing/repaint/composite-and-scale-change-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/fast/repaint/translate-animation-repaint-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/svg/compositing/transform-change-repainting-no-viewBox-repaintRects-expected.txt:
* LayoutTests/platform/mac-wk1/svg/compositing/transform-change-repainting-viewBox-repaintRects-expected.txt:
* LayoutTests/svg/compositing/transform-change-repainting-viewBox-repaintRects-expected.txt:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::~RenderLayer):
(WebCore::RenderLayer::removeSelfFromCompositor):
(WebCore::RenderLayer::recursiveUpdateLayerPositions):
(WebCore::RenderLayer::computeRepaintRects):
(WebCore::RenderLayer::setBackingProviderLayer):
(WebCore::RenderLayer::disconnectFromBackingProviderLayer):
(WebCore::RenderLayer::calculateClipRects const):
(WebCore::outputLayerPositionTreeLegend):
(WebCore::outputLayerPositionTreeRecursive):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::willBeDestroyed):
(WebCore::clearBackingSharingLayerProviders):
(WebCore::RenderLayerBacking::setBackingSharingLayers):
(WebCore::RenderLayerBacking::removeBackingSharingLayer):
(WebCore::RenderLayerBacking::clearBackingSharingLayers):
* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::BackingSharingState::issuePendingRepaints):
(WebCore::RenderLayerCompositor::updateCompositingLayers):
(WebCore::RenderLayerCompositor::computeCompositingRequirements):
(WebCore::RenderLayerCompositor::updateBackingSharingBeforeDescendantTraversal):
(WebCore::RenderLayerCompositor::updateBackingSharingAfterDescendantTraversal):
(WebCore::RenderLayerCompositor::layerGainedCompositedScrollableOverflow):
(WebCore::RenderLayerCompositor::layerStyleChanged):
(WebCore::clearBackingSharingWithinStackingContext):
(WebCore::RenderLayerCompositor::updateExplicitBacking):
(WebCore::RenderLayerCompositor::updateBacking):
(WebCore::RenderLayerCompositor::updateReflectionCompositingState):
(WebCore::RenderLayerCompositor::repaintOnCompositingChange):
(WebCore::RenderLayerCompositor::layerWillBeRemoved):
(WebCore::RenderLayerCompositor::clearBackingForAllLayers):
(WebCore::RenderLayerCompositor::updateLayerCompositingState): Deleted.
* Source/WebCore/rendering/RenderLayerCompositor.h:

Canonical link: https://commits.webkit.org/292627@main
